cli/args/mod.rs

```diff
@@ -664,6 +682,54 @@ impl CliOptions {
     .map(Some)
   }
 
+  /// Returns None for subcommands without a main module
+  pub fn resolve_main_module(
+    &self,
+  ) -> Option<Result<ModuleSpecifier, AnyError>> {
+    match &self.flags.subcommand {
+      DenoSubcommand::Bundle(bundle_flags) => Some(
+        resolve_url_or_path(&bundle_flags.source_file).map_err(AnyError::from),
+      ),
+      DenoSubcommand::Compile(compile_flags) => Some(
+        resolve_url_or_path(&compile_flags.source_file).map_err(AnyError::from),
+      ),
+      DenoSubcommand::Eval(_) => {
+        Some(resolve_url_or_path("./$deno$eval").map_err(AnyError::from))
+      }
+      DenoSubcommand::Repl(_) => {
+        Some(resolve_url_or_path("./$deno$repl.ts").map_err(AnyError::from))
+      }
+      DenoSubcommand::Run(run_flags) => Some(if run_flags.is_stdin() {
+        resolve_url_or_path("./$deno$stdin").map_err(AnyError::from)
+      } else if self.flags.watch.is_some() {
+        resolve_url_or_path(&run_flags.script).map_err(AnyError::from)
+      } else if NpmPackageReqReference::from_str(&run_flags.script).is_ok() {
+        ModuleSpecifier::parse(&run_flags.script).map_err(AnyError::from)
+      } else {
+        resolve_url_or_path(&run_flags.script).map_err(AnyError::from)
+      }),
+      _ => None,
```

Comment:
`resolve_main_module()` is called by `resolve_file_header_overrides()`, which is called by

- `create_graph_and_maybe_check()`
- `ProcState::prepare_module_load()`
- `ProcState::create_graph_loader()`

These calls are unconditional, leaving us with `Option<Result>` because `resolve_main_module()` is called even if there is no main module.